### PR TITLE
TKO-21 -- BE: Implement IDM & User/Admin Authentication (Passport)

### PR DESCRIPTION
**Pull request by MikkalMc**
# Description
'
Added Laravel Passport for oauth token generation and verification
See docs: https://laravel.com/docs/5.8/passport
'
  
# Testing
'tester should be able to hit the following routes to retrieve bearer tokens, register users and revoke bearer tokens: 
'
- /login
- /register
- /logout

# Installation
'Run a composer install and migrate '

### CODE REVIEW

**Review by alecnicolas — COMMENTED** [review]
Works for me 👍

File: app/Http/Controllers/LoginController.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<?php
+namespace App\Http\Controllers;
+
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use App\Contracts\LoginContract;
+use App\User;
+
+class LoginController extends Controller
```
Comment (ChanM0) [inline app/Http/Controllers/LoginController.php:9]: May want to rename to include the idea that logout is in here, couldnt tell by just the name.

File: app/Http/Controllers/LoginController.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    private $loginRetriever;
+
+    public function __construct(LoginContract $loginContract)
+    {
+        $this->loginRetriever = $loginContract;
+    }
+
+    public function login(Request $request)
+    {
+        $request->validate([
```
Comment (ChanM0) [inline app/Http/Controllers/LoginController.php:21]: Same as register should have a Request Object defining the rules and the return messages. Want to make a test for this.

File: app/Services/LoginService.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+use App\Contracts\LoginContract;
+use Carbon\Carbon;
+
+class LoginService implements LoginContract
+{
+    public function login($request)
+    {
+        $credentials = request(['email', 'password']);
+
+        if(!Auth::attempt($credentials))
```
Comment (ChanM0) [inline app/Services/LoginService.php:15]: middle wear around api route does this already.

File: app/Services/LoginService.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            return response()->json([
+                'message' => 'Unauthorized'
+            ], 401);
+
+            $user = $request->user();
```
Comment (ChanM0) [inline app/Services/LoginService.php:20]: Use Auth::user();

File: app/Http/Controllers/LoginController.php (hunk 2 above)
Comment (MikkalMc) [reply app/Http/Controllers/LoginController.php:21]: task made to address this

**Review by alecnicolas — APPROVED** [review]
idk what changed but it still works for me so

File: app/Services/RegisterService.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+use function Opis\Closure\unserialize;
+use function Opis\Closure\serialize;
+
+
+class RegisterService implements RegisterContract
+{
+    public function register($request)
+    {
+        try {
+            $name = (string)$request['name'];
```
Comment (luisjg) [inline app/Services/RegisterService.php:18]: Aren't these strings already?

File: app/Services/RegisterService.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            $email = (string)$request['email'];
+            $password = (string)$request['password'];
+        } catch (\Exception $e) {
+            return ['message_error' => 'User was not successfully created.'];
+        }
+
+        //TODO: add organization destinction
+        try {
```
Comment (luisjg) [inline app/Services/RegisterService.php]: There's a createOrFail function or you can surround this in a database transaction -> https://laravel.com/docs/5.8/database#database-transactions

File: resources/js/api/Auth.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// AUTH API
+//
+
+import axios from 'axios';
+
+axios.get('/oauth/clients')
+    .then(response => {
+        console.log(response.data);
```
Comment (luisjg) [inline resources/js/api/Auth.js:8]: Console logs....

**Review by luisjg — CHANGES_REQUESTED** [review]
See comments.

**Review by luisjg — APPROVED** [review]
👍

**Pull request merged**